src/coreclr/vm/assemblynative.cpp

```diff
@@ -1429,3 +1430,313 @@ extern "C" BOOL QCALLTYPE AssemblyNative_IsApplyUpdateSupported()
 
     return result;
 }
+
+namespace
+{
+    LPCSTR TypeMapAssemblyTargetAttributeName = "System.Runtime.InteropServices.TypeMapAssemblyTargetAttribute`1";
+    LPCSTR TypeMapAttributeName = "System.Runtime.InteropServices.TypeMapAttribute`1";
+    LPCSTR TypeMapAssociationAttributeName = "System.Runtime.InteropServices.TypeMapAssociationAttribute`1";
+
+    bool IsTypeSpecForTypeMapGroup(
+        MethodTable* groupTypeMT,
+        Assembly* pAssembly,
+        mdToken typeSpec)
+    {
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(groupTypeMT != NULL);
+        _ASSERTE(pAssembly != NULL);
+        _ASSERTE(TypeFromToken(typeSpec) == mdtTypeSpec);
+
+        IMDInternalImport* pImport = pAssembly->GetMDImport();
+
+        PCCOR_SIGNATURE sig;
+        ULONG sigLen;
+        IfFailThrow(pImport->GetTypeSpecFromToken(typeSpec, &sig, &sigLen));
+
+        SigPointer sigptr{ sig, sigLen };
+
+        // Processing TypeSpec. See ECMA-225, II.23.2.14 TypeSpec
+        CorElementType type;
+        IfFailThrow(sigptr.GetElemType(&type));
+        if (type != ELEMENT_TYPE_GENERICINST)
+            return false;
+
+        IfFailThrow(sigptr.GetElemType(&type));
+        if (type != ELEMENT_TYPE_CLASS) // All TypeMap attributes are classes.
+            return false;
+
+        // TypeMap attribute type token
+        IfFailThrow(sigptr.GetToken(NULL));
+
+        uint32_t genericArgCount;
+        IfFailThrow(sigptr.GetData(&genericArgCount));
+        if (genericArgCount != 1) // All TypeMap attributes have a single generic parameter.
+            return false;
+
+        IfFailThrow(sigptr.GetElemType(&type));
+        switch (type)
+        {
+        case ELEMENT_TYPE_OBJECT:
+            return groupTypeMT == g_pObjectClass;
+        case ELEMENT_TYPE_STRING:
+            return groupTypeMT == g_pStringClass;
+        case ELEMENT_TYPE_ARRAY:
+            return false;
+        default:
+            break;
+        }
+
+        CorElementType groupElementType = groupTypeMT->GetSignatureCorElementType();
+        if (type != groupElementType)
+            return false;
+
+        if (type != ELEMENT_TYPE_CLASS && type != ELEMENT_TYPE_VALUETYPE)
+        {
+            // No token to compare if the element type is not a class or value type.
+            return true;
+        }
```

Comment:
Open to other opinions about this logic. I used `Guid`, `string` and `object` for some testing so I handled these cases. However, I've no idea how to think about instantiated generic types, open generics, etc.

Reply:
I think we should reject open generic types for now. I’m not sure the semantics are well defined in the presence of trimming (particularly in the case of the “tested” type in the native->managed map), and we can always add support later.

Reply:
Fixed.